File: recursao/multiplica_por_dois.c
/**
 * Função que multiplica todos os elementos de um vetor por 2
 * de forma recursiva.
*/



void M2(int vetor[], int n) {
    if (n == 0) {
        return;  // cabô
    }
    //else
    vetor[n-1] = vetor[n-1] * 2;
    M2(vetor, n-1);
}


int main() {

    int n = 5;
    int v1[] = {10, 30, 50, 70, 90};

    M2(v1, n);

    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d, ", v1[i]);
    }
    printf("]");

    return 0;
}

File: recursao/multiplica_por_dois.py
"""
Exemplo usando recursão em cauda, típico de linguagens funcionais.
Em algumas linguagens, especialmente as puramente funcionais,
esse tipo de recursão é otimizado de forma que o compilador o
transforma em uma iteração sobre a lista. Dessa forma,
não ocorre a desvantagem da recursão, que é o uso excessivo de
memória e overflow de pilha.

Em outras linguagens, como C e Python, não há essa otimização,
logo, é mais recomendado que se utilize a versão iterativa.
"""

def M2(lista):
    if not lista:  #vetor vazio
        return []
    #else:
    primeiro_mod = lista[0] * 2
    return [primeiro_mod] + M2(lista[1:])
    # explicação: concatena-se o primeiro modificado
    # à aplicação de M2 ao resto (cauda) da lista


# Testes:
v1 = [10, 30, 50, 70, 90]
v2 = M2(v1)
print(v2)